ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl.hs

```diff
@@ -315,10 +336,490 @@ cdbStreamBlocks CDB{..} = implStreamBlocks cdbImmDB cdbVolDB
   been initialized
 -------------------------------------------------------------------------------}
 
-implStreamBlocks :: ImmDB m blk
+-- | Stream blocks
+--
+-- TODO handle corruption and recovery
+-- TODO test that EBBs are handled correctly
+--
+-- = Start & end point
+--
+-- The start point can either be in the ImmutableDB (on our chain) or in the
+-- VolatileDB (on our chain or on a recent fork). We first check whether it is
+-- in the VolatileDB, if not, we check if it is in the ImmutableDB (see
+-- \"Garbage collection\" for why this order is important). Similarly for the
+-- end point.
+--
+-- If a bound can't be found in the ChainDB, an 'UnknownRange' error is
+-- returned.
+--
+-- When the bounds are nonsensical, e.g.,
+-- > StreamFromExclusive (Point { pointSlot = SlotNo 3 , .. }
+-- > StreamToInclusive   (Point { pointSlot = SlotNo 3 , .. }
+-- An 'InvalidIteratorRange' exception is thrown.
+--
+-- = Paths of blocks
+--
+-- To stream blocks from the ImmutableDB we can simply use the iterators
+-- offered by the ImmutableDB.
+--
+-- To stream blocks from the VolatileDB we have to construct a path of block
+-- hashes backwards through the VolatileDB, starting from the end point using
+-- 'getPredecessor' until we get to the start point, genesis, or we get to a
+-- block that is not in the VolatileDB. Then, for each hash in the path, we
+-- can ask the VolatileDB for the corresponding block.
+--
+-- If the path through the VolatileDB is incomplete, we will first have to
+-- stream blocks from the ImmutableDB and then switch to the path through the
+-- VolatileDB. We only allow the tip of the ImmutableDB to be the switchover
+-- point between the two DBs. In other words, the incomplete path through the
+-- VolatileDB must fit onto the tip of the ImmutableDB. If it doesn't fit on
```

Comment:
When initialising the iterator. It might not be the case anymore during streaming.